File: src/types/duplicator.ts
/** Types mirroring the backend Exercise Duplicator schemas. */

export type VizType =
  | "mafs_function"
  | "mafs_parametric"
  | "plotly_3d"
  | "plotly_complex"
  | "matter_simulation"
  | "force_diagram";

export interface VizBlock {
  type: VizType;
  expression?: string;
  domain?: [number, number];
  params?: Record<string, unknown>;
}

export type ExerciseSessionStatus = "extracting" | "ready" | "error";
export type ExerciseStatus = "pending" | "searching" | "done" | "error";

export interface DuplicateResult {
  id: number;
  source_url: string | null;
  problem_text: string;
  viz: VizBlock | null;
  difficulty_score: number | null;
}

export interface ExtractedExercise {
  id: number;
  order_index: number;
  raw_text: string;
  topic: string;
  subtopic: string | null;
  difficulty_signals: string[];
  viz: VizBlock | null;
  status: ExerciseStatus;
  results: DuplicateResult[];
}

export interface ExerciseSession {
  id: number;
  document_hash: string;
  year_level: number;
  subject_hint: string | null;
  status: ExerciseSessionStatus;
  exercises: ExtractedExercise[];
}

File: backend/schemas/duplicator.py
"""Schemas for the Exercise Duplicator API (sessions + exercises + results).

``VizBlock`` mirrors the renderer contract the frontend reads (``viz.type`` routes
to a renderer). It's kept permissive — a model may emit any of the documented
types plus the renderer-specific ``expression`` / ``domain`` / ``params`` — so the
backend never rejects an otherwise-usable exercise over a viz field.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict

from backend.models.enums import ExerciseSessionStatus, ExerciseStatus


class DuplicateResultOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_url: str | None
    problem_text: str
    viz: dict[str, Any] | None
    difficulty_score: float | None


class ExtractedExerciseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    order_index: int
    raw_text: str
    topic: str
    subtopic: str | None
    difficulty_signals: list[str]
    viz: dict[str, Any] | None
    status: ExerciseStatus
    results: list[DuplicateResultOut] = []


class ExerciseSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_hash: str
    year_level: int
    subject_hint: str | None
    status: ExerciseSessionStatus
    exercises: list[ExtractedExerciseOut] = []
